### DISCUSSION TITLE
Praise: Update firmware worked on tricky device!

### DISCUSSION THREAD

**Discussion by LordMike** [zwave-js/zwave-js#1463] [General]
I acquired a firmware from a manufacturer to fix an issue, and I actually got an .otz file, which is more bare-bones than other companies provide, like aeotec's exe files etc.

I was on the edge on using zwave-js for it, no guarantees and all.. but doing a few upgrades with the silabs pc controller, I kinda got the feeling that there were so many checks, checksums and so on .. and that the node wouldn't complete an upgrade unless it got all parts intact.. (heres to hoping).

One of my devices just refused to update.. So hey - I tried zwave-js (through zwavejs2mqtt). Pick node, do upgrade, pick file.. and off it went. It went immensely slow (bad routes?), approx 22 minutes.. probably why silabs pc controller failed.. But it worked in the  end :O

And now the device reports the new firmware. 

🎉

**Comment by hanskroner** [comment]
Hooray for interoperability!!! 😝

> I was on the edge on using zwave-js for it, no guarantees and all.. but doing a few upgrades with the silabs pc controller, I kinda got the feeling that there were so many checks, checksums and so on .. and that the node wouldn't complete an upgrade unless it got all parts intact.. (heres to hoping).

A lot of work has gone into making this process resilient on the Z-Wave side of things. Nobody will go as far as to saying it's perfect, but the checks and balances you mention - and more - are in place.

> One of my devices just refused to update.. So hey - I tried zwave-js (through zwavejs2mqtt). Pick node, do upgrade, pick file.. and off it went. It went immensely slow (bad routes?), approx 22 minutes.. probably why silabs pc controller failed.. But it worked in the end :O

Depending on the size of the firmware file, updates can take "a while." 100kbps and limited frame sizes doesn't give you a lot of bandwidth. There's also the overhead from the Command Classes used to transport the data - particularly when using Security, which not only limits the number of bytes in a frame available for the payload, but might also have `NONCE_REPORT` frames overhead. It all takes even longer if frames need to be routed.

The PC Controller is pretty good at handling this though. Anecdotally, I've used it to test OTA of battery devices that request a small number of chunks of the new firmware every couple of hours, so the process takes **days**.

**Reply by LordMike** [reply]
I have no idea why PC Controller couldn't.. It reports each frame it sends (also in the message log), and it never passed frame 1. Eventually it times out after a short while (minutes), and then aborts.. The PC controller had no issue sending other messages, like getting the current firmware version or doing a "node info".

But lets not dwell on that :)
